include/dlaf/util_matrix.h

```diff
@@ -21,149 +21,189 @@ constexpr double M_PI = 3.141592;
-            function + ": " + "The matrices " + mat_a_name + " and " + mat_b_name +
-            " are not right multipliable (size of matrix A not equal to that of matrix B).");
-      }
-      if (mat_a.blockSize().rows() != mat_b.blockSize().cols()) {
-        throw std::invalid_argument(
-            function + ": " + "The matrices " + mat_a_name + " and " + mat_b_name +
-            " are not right multipliable (blocksize of matrix A not equal to that of matrix B).");
-      }
-    }
-    else if (op == blas::Op::Trans || op == blas::Op::ConjTrans) {
-      if (mat_a.nrTiles().cols() != mat_b.nrTiles().cols()) {
-        throw std::invalid_argument(
-            function + ": " + "The matrices " + mat_a_name + " and " + mat_b_name +
-            " are not right multipliable (rows of matrix A not equal to cols of matrix B).");
-      }
-      if (mat_a.size().cols() != mat_b.size().cols()) {
-        throw std::invalid_argument(
-            function + ": " + "The matrices " + mat_a_name + " and " + mat_b_name +
-            " are not right multipliable (size of matrix A not equal to that of matrix B).");
-      }
-      if (mat_a.blockSize().cols() != mat_b.blockSize().cols()) {
-        throw std::invalid_argument(
-            function + ": " + "The matrices " + mat_a_name + " and " + mat_b_name +
-            " are not right multipliable (blocksize of matrix A not equal to that of matrix B).");
-      }
+/// When the assertion is enabled, terminates the program with an error message if matrix is not on distributed
+/// according to the given communicator grid. This assertion is enabled when **DLAF_ASSERT_ENABLE** is ON.
+#define DLAF_ASSERT_DISTRIBUTED_ON_GRID(grid, matrix)                                                \
+  DLAF_ASSERT(((matrix.distribution().commGridSize() == grid.size()) &&                              \
+               (matrix.distribution().rankIndex() == grid.rank())),                                  \
+              "The matrix ", #matrix, " (rank: ", matrix.distribution().rankIndex(),                 \
+              ", grid size: ", matrix.distribution().commGridSize().rows(), "x",                     \
+              matrix.distribution().commGridSize().cols(),                                           \
+              ") is not distributed according to the communicator grid ", #grid,                     \
+              " (rank: ", grid.rank(), ", grid size: ", grid.size().rows(), "x", grid.size().cols(), \
+              ").")
+
+template <class MatrixConst, class Matrix, class Mat, class Location>
+void assertMultipliableMatrices(const MatrixConst& mat_a, const Matrix& mat_b, const Mat& mat_c,
+                                const blas::Op opA, const blas::Op opB, const Location location,
+                                std::string mat_a_name, std::string mat_b_name, std::string mat_c_name) {
+  auto get_k = [](const auto& size, const blas::Op op, bool first_member) -> decltype(size.rows()) {
```

Comment:
I would go for a much simpler version:
```
{
  auto rows = [](const auto& size, const blas::Op op) -> decltype(size.rows()) {
    switch (op) {
      case blas::Op::NoTrans:
        return size.rows();
      case blas::Op::Trans:
      case blas::Op::ConjTrans:
        return size.cols();
      default:
        return {};
    }
  }
  auto cols = [](const auto& size, const blas::Op op) -> decltype(size.cols()) {
    <similar as above>
  }

  DLAF_ASSERT_WITH_ORIGIN(location, rows(mat_a, opA) == mat_c.rows() && cols(mat_a, opA) == rows(mat_b, opB) && cols(mat_b, opB) == mat_c.cols(), <error_message> );
}
```
For the error message I suggest:
Size mismatch: <size_opA_mat_a> x <size_opB_mat_B> -> <size_mat_c> cannot be performed.

Reply:
Look great. Way more simple. 
Do we need to check also `blockSize` and `nrTiles` "sizes"?

Reply:
Yes the same has to be done for the blockSize (nrTiles follows from size and blocksize).
A second assert can be added with `blockSize()` instead of `size()`.
For the error message I suggest:
Block-size mismatch: <block_size_opA_mat_a> x <block_size_opB_mat_B> -> <block_size_mat_c> cannot be performed.

Reply:
Thanks.